### PR TITLE
Catch up with Firedrake updates

### LINKED ISSUE #61: Adjoint issues after updating firedrake / leaving pyroteus
Hey @jwallwork23 and @stephankramer, could you please help me debug the errors I get when solving the adjoint on the new firedrake installation? Everything worked well on my previous installation (Using pyroteus and Joe's parmmg-metric-based firedrake branch, which I last updated some 6 months ago). Now I am on the firedrake master branch and using animate and goalie.

I get this error in my glacier experiments, but also when running the demos (e.g. burgers-ee.py and point_discharge2d-goal_oriented.py): `ValueError: Looks like no solves were written to tape! Does the solution depend on the initial condition?`. Here's the debug and traceback from burgers_ee.py:

```
WARNING GoalOrientedMeshSeq: Tape has no blocks!
WARNING GoalOrientedMeshSeq: Tape has no blocks!
WARNING GoalOrientedMeshSeq: No block has an adjoint solution. Has the adjoint equation been solved?
Traceback (most recent call last):
  File "/home/ubuntu/software/firedrake-sep23/src/goalie/demos/burgers_ee.py", line 139, in <module>
    solutions, indicators = mesh_seq.indicate_errors(
  File "petsc4py/PETSc/Log.pyx", line 115, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "petsc4py/PETSc/Log.pyx", line 116, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "/home/ubuntu/software/firedrake-sep23/src/goalie/goalie/go_mesh_seq.py", line 141, in indicate_errors
    sols = self.solve_adjoint(**adj_kwargs)
  File "petsc4py/PETSc/Log.pyx", line 115, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "petsc4py/PETSc/Log.pyx", line 116, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "/home/ubuntu/software/firedrake-sep23/src/goalie/goalie/adjoint.py", line 323, in solve_adjoint
    raise ValueError(
ValueError: Looks like no solves were written to tape! Does the solution depend on the initial condition?
```

I made some importing changes inside goalie. Here they are:
1. goalie currently imports `firedrake_adjoint` which gave me a warning `FutureWarning: The firedrake_adjoint module is deprecated. Instead, use the firedrake.adjoint module and explicitly start taping by calling continue_annotation().` So I changed all such imports to `firedrake.adjoint`, but I _**did not**_ call this `continue_annotation()`. I wasn't sure where to call this. Before I did that, by still importing `firedrake_adjoint`, I would get the following error:
```
Traceback (most recent call last):
  File "/home/ubuntu/software/firedrake-sep23/src/PyOP2/pyop2/caching.py", line 205, in __new__
    return cls._cache_lookup(key)
  File "/home/ubuntu/software/firedrake-sep23/src/PyOP2/pyop2/caching.py", line 213, in _cache_lookup
    return cls._cache[key]
KeyError: (<class 'pyop2.global_kernel.GlobalKernel'>, ('0267ee8ad062db3cfcadfaa6e6454f87', (<Access.WRITE: 2>, <Access.READ: 1>, <Access.READ: 1>, <Access.READ: 1>, <Acc
ess.READ: 1>, <Access.READ: 1>, <Access.READ: 1>, <Access.READ: 1>), (np:dtype('float64'), np:dtype('float64'), np:dtype('float64'), np:dtype('float64'), np:dtype('float6
4'), np:dtype('float64'), np:dtype('float64'))), ('extruded', False), ('extruded_periodic', False), ('constant_layers', False), ('subset', False), 4, (<class 'pyop2.globa
l_kernel.MatKernelArg'>, (((1, 1),),), ((<class 'pyop2.global_kernel.MapKernelArg'>, 3, None, None), (<class 'pyop2.global_kernel.MapKernelArg'>, 3, None, None)), False),
 (<class 'pyop2.global_kernel.DatKernelArg'>, (1,), (<class 'pyop2.global_kernel.MapKernelArg'>, 3, None, None), None), (<class 'pyop2.global_kernel.DatKernelArg'>, (1,),
 (<class 'pyop2.global_kernel.MapKernelArg'>, 3, None, None), None), (<class 'pyop2.global_kernel.DatKernelArg'>, (1,), (<class 'pyop2.global_kernel.MapKernelArg'>, 3, No
ne, None), None), (<class 'pyop2.global_kernel.GlobalKernelArg'>, (1,)), (<class 'pyop2.global_kernel.GlobalKernelArg'>, (1,)), (<class 'pyop2.global_kernel.GlobalKernelA
rg'>, (1,)), (<class 'pyop2.global_kernel.GlobalKernelArg'>, (1,)), 0, 0, 0, 0, 0)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/data/meshadapt/Ice1-27sep2023/./timedep_goal.py", line 445, in <module>
    msq.fixed_point_iteration(adaptor)
  File "petsc4py/PETSc/Log.pyx", line 115, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "petsc4py/PETSc/Log.pyx", line 116, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "/home/ubuntu/software/firedrake-sep23/src/goalie/goalie/go_mesh_seq.py", line 283, in fixed_point_iteration 
    sols, indicators = self.indicate_errors(
  File "petsc4py/PETSc/Log.pyx", line 115, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "petsc4py/PETSc/Log.pyx", line 116, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "/home/ubuntu/software/firedrake-sep23/src/goalie/goalie/go_mesh_seq.py", line 141, in indicate_errors
    sols = self.solve_adjoint(**adj_kwargs)
  File "petsc4py/PETSc/Log.pyx", line 115, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "petsc4py/PETSc/Log.pyx", line 116, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "/home/ubuntu/software/firedrake-sep23/src/goalie/goalie/adjoint.py", line 315, in solve_adjoint
    tape.evaluate_adj(markings=True)
  File "/home/ubuntu/software/firedrake-sep23/src/pyadjoint/pyadjoint/tape.py", line 228, in evaluate_adj
    self._blocks[i].evaluate_adj(markings=markings)
  File "/home/ubuntu/software/firedrake-sep23/src/pyadjoint/pyadjoint/tape.py", line 105, in wrapper
    return function(*args, **kwargs)
  File "/home/ubuntu/software/firedrake-sep23/src/pyadjoint/pyadjoint/block.py", line 131, in evaluate_adj
    adj_output = self.evaluate_adj_component(inputs,
  File "/home/ubuntu/software/firedrake-sep23/src/firedrake/firedrake/adjoint_utils/blocks/function.py", line 335, in evaluate_adj_component
    return self.backend.Interpolator(dJdm, self.V).interpolate(adj_inputs[0], output=output, transpose=True)
  File "/home/ubuntu/software/firedrake-sep23/src/firedrake/firedrake/interpolation.py", line 621, in __init__
    self.callable, arguments = make_interpolator(expr, V, subset, access, bcs=bcs)
  File "petsc4py/PETSc/Log.pyx", line 115, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "petsc4py/PETSc/Log.pyx", line 116, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "/home/ubuntu/software/firedrake-sep23/src/firedrake/firedrake/interpolation.py", line 801, in make_interpolator
    loops.extend(_interpolator(V, tensor, expr, subset, arguments, access, bcs=bcs))
  File "<decorator-gen-31>", line 2, in _interpolator
  File "/home/ubuntu/software/firedrake-sep23/src/firedrake/firedrake/utils.py", line 75, in wrapper
    return f(*args, **kwargs)
  File "/home/ubuntu/software/firedrake-sep23/src/firedrake/firedrake/interpolation.py", line 985, in _interpolator 
    parloop = op2.ParLoop(*parloop_args)
  File "/home/ubuntu/software/firedrake-sep23/src/PyOP2/pyop2/parloop.py", line 606, in ParLoop
    return LegacyParloop(*args, **kwargs)
  File "/home/ubuntu/software/firedrake-sep23/src/PyOP2/pyop2/parloop.py", line 635, in LegacyParloop
    global_knl = GlobalKernel(local_knl, global_knl_args,
  File "/home/ubuntu/software/firedrake-sep23/src/PyOP2/pyop2/caching.py", line 207, in __new__
    obj = make_obj()
  File "/home/ubuntu/software/firedrake-sep23/src/PyOP2/pyop2/caching.py", line 197, in make_obj
    obj.__init__(*args, **kwargs)
  File "/home/ubuntu/software/firedrake-sep23/src/PyOP2/pyop2/global_kernel.py", line 287, in __init__
    for larg, garg in zip(local_kernel.arguments, arguments)
  File "/home/ubuntu/software/firedrake-sep23/src/PyOP2/pyop2/local_kernel.py", line 144, in arguments
    assert len(self.accesses) == len(self.dtypes)
AssertionError
```
2. Changed import in L6 in goalie.mesh_seq.py from `from firedrake.adjoint.solving import get_solve_blocks` to `from firedrake.adjoint_utils.solving import get_solve_blocks` (because `firedrake.adjoint.solving` doesn't exist on the master branch). This seems like an obvious issue based on the warning above `WARNING GoalOrientedMeshSeq: Tape has no blocks!`
3. Other small import changes which I don't think are important here. E.g. in goalie.options.py changed `from firedrake.meshadapt import RiemannianMetric` to `from animate.adapt import RiemannianMetric`.

### PR DESCRIPTION
**Pull request by jwallwork23**
Closes #61.
Take 2 of #62.

Okay there ended up being quite a few ways we had got out of date with Firedrake... apologies for the mega PR.

Summary:
* Fix import paths.
* Use `Cofunction` and `FunctionSpace.dual()`.
* Update `norm` and `errornorm` to accept `Cofunction`s.
* Update `project` to accept `Cofunction`s.
* To enable the above two, I needed to introduce `function2cofunction` and `cofunction2function` utilities.
* Use `y = Function(FunctionSpace(mesh, "R", 0)).assign(x)` rather than `y = Constant(x)`.
* Update tests accordingly.

### CODE REVIEW

**Comment by jwallwork23** [comment]
Oh and the `Cofunction` changes also made me realise a problem with the error estimation code... The flux terms were summarised by a `Function` which came from a linear solve, whereas the cell terms came from assembling a 1-form, which is a `Cofunction`. After changing the latter to reflect the former, I found a load of bugs in the error estimation testing. Hopefully, it gives better results now :)

File: goalie/error_estimation.py [hunk 1]
```diff
@@ -59,7 +58,15 @@ def form2indicator(F: ufl.form.Form) -> Function:
...
-        indicator += firedrake.assemble(cell_terms)
+        cell_contrib = Function(P0)
+        sp = {
+            "snes_type": "ksponly",
+            "ksp_type": "preonly",
+            "pc_type": "lu",
+            "pc_factor_mat_solver_type": "mumps",
+        }
+        firedrake.solve(mass_term == cell_terms, cell_contrib, solver_parameters=sp)
+        indicator += cell_contrib
```
Comment (jwallwork23) [inline goalie/error_estimation.py:69]: Question that came up in the meeting: what's the difference between this and the original, other than the type? Is this system also trivial? If so, we could solve it using a simple application of a Jacobi preconditioner, as above.

Comment (stephankramer) [reply goalie/error_estimation.py:69]: Sorry, probably misunderstood yesterday: this is indeed changed now: before you add in a cell-integration of the integrand, but now you're adding in a cell-average (i.e. divided by cell area - or equivalently this is just a projection into P0). Don't know which one is correct

File: goalie/interpolation.py [hunk 2]
```diff
@@ -1,72 +1,64 @@
...
-        raise NotImplementedError("Can only currently project Functions.")  # TODO
-    Vs = source.function_space()
-    if isinstance(target_space, firedrake.Function):
+    if not isinstance(source, (firedrake.Function, firedrake.Cofunction)):
+        raise NotImplementedError(
+            "Can only currently project Functions and Cofunctions."
+        )
+
+    # If the input is a Cofunction then record this, map to a Function for the
+    # projection and then map back to a Cofunction afterwards
```
Comment (stephankramer) [inline goalie/interpolation.py:33]: I was worried when I read this comment - as it seems to suggest you convert to a function, then do a normal projection and then convert back - but you are actually doing an adjoint projection

File: goalie/interpolation.py [hunk 3]
```diff
@@ -1,72 +1,64 @@
...
+            "Second argument must be a FunctionSpace, Function, or Cofunction."
         )
 
-    # Apply projector
-    return (_project_adjoint if adjoint else _project)(source, target, **kwargs)
+    # Choose appropriate transfer method
+    if source.function_space() == target.function_space():
+        target.assign(source)
+    elif adjoint:
+        target = _project_adjoint(source, target, **kwargs)
```
Comment (stephankramer) [inline goalie/interpolation.py:54]: Would it not make more sense to have `_projection_adjoint` only take a cofunctions as source and target (it only pulls out the vector anyway so it wouldn't need any conversion). I don't think doing an adjoint projection on normal functions really makes sense...

**Review by stephankramer — APPROVED** [review]

File: goalie/interpolation.py (hunk 2 above)
Comment (jwallwork23) [reply goalie/interpolation.py]: Reworded in 1b825f0.

File: goalie/interpolation.py (hunk 3 above)
Comment (jwallwork23) [reply goalie/interpolation.py]: Good point! Refactored in 1b825f0.

File: goalie/error_estimation.py (hunk 1 above)
Comment (jwallwork23) [reply goalie/error_estimation.py]: Ah yes, good point. Considering this a bit more, I think this is the right thing to do in f55dce1 - multiply throughout the RHS with `CellSize(mesh)`.

**Comment by jwallwork23** [comment]
I'm gonna go ahead and get this merged as I'm confident it meets the requests. Thanks again for the review @stephankramer!

**Comment by ddundo** [comment]
> * Use `y = Function(FunctionSpace(mesh, "R", 0)).assign(x)` rather than `y = Constant(x)`.

@jwallwork23 just a tiny comment here - if we pass a `domain` argument to `Constant`, i.e. `y = Constant(x, domain=mesh)`, this is syntactic sugar for `y = Function(FunctionSpace(mesh, "R", 0)).assign(x)`. Apparently this is because anything which is  a variable needs to know what mesh it's on for parallel safety reasons. I see that you're using the latter in the demos so I wasn't sure if you prefer it for clarity or maybe weren't aware of the first one :)

**Pull request merged**